create_database/pubtator-to-postgres.py

```diff
@@ -0,0 +1,392 @@
+    doc='.//document',
+    text='.//passage/text/text()',
+    id='.//id/text()', tag_filter=set(filter_df['pubmed_id']))
+
+
+# In[ ]:
+
+
+dg_tagger = Tagger(grouped)
+
+
+# In[ ]:
+
+
+corpus_parser = CorpusParser(fn=dg_tagger.tag)
+document_chunk = []
+
+for document in tqdm.tqdm(xml_parser.generate()):
+    
+    document_chunk.append(document)
+
+    # chunk the data because snorkel cannot 
+    # scale properly
+    if len(document_chunk) >= 5e4:
+        corpus_parser.apply(document_chunk, parallelism=5, clear=False)
+        document_chunk = []
+    
+# If generator exhausts and there are still
+# document to parse
+if len(document_chunk) > 0:
+    corpus_parser.apply(data, parallelism=5, clear=False)
+    document_chunk = []
+
+
+# # Get each candidate relation
+
+# After parsing the above abstracts, the next step in this pipeline is to extract candidates from all the tagged sentences. A candidate is considered a candidate if two mentions occur in the same sentence. For this pilot study, we are only considering the follow candidate relationships: Disease-Gene, Gene-Gene, Compound-Gene, Compound-Disease. In conjunction with extracting candidates, this part of the pipeline also stratifies each sentence into three different categories: Train (70%), Dev (20%), and Test (10%). These set categories will be used in subsequent notebooks ([3](3.data-gen-model.ipynb), [4](4.data-disc-model.ipynb), [5](5.data-analysis.ipynb)) for training and testing the machine learning algorithms.
+
+# In[ ]:
+
+
+chunk_size = 2e5
+
+
+# In[ ]:
+
+
+def insert_cand_to_db(extractor, sentences):
+    for split, sens in enumerate(sentences):
+        extractor.apply(sens, split=split, parallelism=5, clear=False)
+
+
+# In[ ]:
+
+
+def print_candidates(context_class, edge):
+    for i, label in enumerate(["Train", "Dev", "Test"]):
+        cand_len = session.query(context_class).filter(context_class.split == i).count()
+        print("Number of Candidates for {} edge and {} set: {}".format(edge, label, cand_len))
+
+
+# In[ ]:
+
+
+#This specifies the type of candidates to extract
+DiseaseGene = candidate_subclass('DiseaseGene', ['Disease', 'Gene'])
+dge = PretaggedCandidateExtractor(DiseaseGene, ['Disease', 'Gene'])
+
+GeneGene = candidate_subclass('GeneGene', ['Gene1', 'Gene2'])
+gge = PretaggedCandidateExtractor(GeneGene, ['Gene', 'Gene'])
+
+CompoundGene = candidate_subclass('CompoundGene', ['Compound', 'Gene'])
+cge = PretaggedCandidateExtractor(CompoundGene, ['Compound', 'Gene'])
+
+CompoundDisease = candidate_subclass('CompoundDisease', ['Compound','Disease'])
+cde = PretaggedCandidateExtractor(CompoundDisease, ['Compound', 'Disease'])
+
+
+# In[ ]:
+
+
+# set the seed for reproduction
+np.random.seed(100)
+total_sentences = session.execute("select count(*) from sentence").fetchone()[0]
+
+
+# In[ ]:
+
+
+category_list = np.random.choice([0,1,2], total_sentences, p=[0.7,0.2,0.1])
+
+
+# In[ ]:
+
+
+# Divide the sentences into train, dev and test sets
```

Comment:
So you are not grouping sentences based on document?  (i.e. all the sentences from document A are not in training but can be split across training and testing?)  I wonder if this will make a difference?  I feel like you might lose some context if they are not grouped, but I also understand wanting to make sure each group is similar in what it has seen

Reply:
Right each sentence is considered its own individual entity. Also, you are correct. The biggest issue with this project was that I lost contextual information, which made sentence prediction extremely difficult (e.g. ambiguous sentences). Something to keep in mind when I move towards full text.